### ISSUE TITLE
Cloning Linux kernel code under Windows

### ISSUE THREAD

**Issue by mfide** [git-for-windows/git#4005]
I use git version 2.37.1.windows.1 on my Windows 10 Pro 64 bit machine with a NTFS partition.
I want to clone some Linux kernel code from git://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git

```
git.exe clone --progress --recursive -v "git://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git" "C:\_Projects\linux-stable"
Cloning into 'C:\_Projects\linux-stable'...
Looking up git.kernel.org ... done.
Connecting to git.kernel.org (port 9418) ... 145.40.68.75 done.
remote: Enumerating objects: 327, done.
remote: Counting objects: 100% (327/327), done.
remote: Compressing objects: 100% (184/184), done.
remote: Total 10739706 (delta 182), reused 203 (delta 143), pack-reused 10739379
Receiving objects: 100% (10739706/10739706), 4.12 GiB | 22.28 MiB/s, done.
Resolving deltas: 100% (8585326/8585326), done.
Checking objects: 100% (33554432/33554432), done.
error: invalid path 'drivers/gpu/drm/nouveau/nvkm/subdev/i2c/aux.c'
fatal: unable to checkout working tree
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry with 'git restore --source=HEAD :/'

git did not exit cleanly (exit code 128) (768672 ms @ 30/08/2022 11:08:03)
```

Then I added following lines in to **.git/config** and try to checkout to master again:
```
[core]
    longpaths = true
    sparsecheckout = true
    protectNTFS = false
```

and then get following errors:
```
git.exe checkout -f master --
error: unable to create file drivers/gpu/drm/nouveau/nvkm/subdev/i2c/aux.c: No such file or directory
error: unable to create file drivers/gpu/drm/nouveau/nvkm/subdev/i2c/aux.h: No such file or directory
error: unable to create file include/soc/arc/aux.h: No such file or directory
Your branch is up to date with 'origin/master'.
Already on 'master'


git did not exit cleanly (exit code 1) (157640 ms @ 30/08/2022 11:13:47)
```

I know that `aux` is one of reserved word in Windows file system, but I wonder if git can manage the file names which are reserved automatically in a similar way they do with CR and LF? (for example it can replace the file name during cloning and committing)

**Comment by dscho** [comment]
Related: https://github.com/git-for-windows/git/issues/2803.

The broader problem is not actually specific to Windows because it will require a change in core Git if we want to allow checking out files under a different name than they were recorded.

Although that might not go over well: If you clone a repository such as the Linux kernel's, there is more than just Git that assumes that the file names are on disk as they were committed, e.g. the build system.

In any case, you probably will want to start a discussion on [the Git mailing list](mailto:git@vger.kernel.org) (send plain-text messages, HTML messages are dropped silently).

**Comment by mfide** [comment]
But interestingly on my desktop, right click -> New -> Text Document and relace the file name as aux.c or aux.h works fine. More than that, I can edit the file with Notepad++ and save back without any restriction.
[https://i.ibb.co/CsJxHcG/aux-issue.png](https://i.ibb.co/CsJxHcG/aux-issue.png)

I didn't get what is the issue with git then.

**Comment by dscho** [comment]
> But interestingly on my win11 desktop, right click -> New -> Text Document and replacing the file name as aux.c or aux.h works fine.

"Works on my machine" is not really adequate if you run a project that's used by millions of users.

**Comment by mfide** [comment]
> "Works on my machine" is not really adequate if you run a project that's used by millions of users.

You are right, I confirmed that on some other Win11 machines as well, obviously Microsoft had some improvement on the topic since Win10, but I couldn't find official statement for it. On the other side, case sensitive files and folders were already supported through a special flag which can be set for a folder via fsutil. I believed that with all those updates, windows file system is fully compatible with Linux Kernel Source code. I will do the further tests and post the results here.

**Comment by Okeanos** [comment]
You just ran into some of the magic that is the [Windows file naming convention](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions). Specifically:


> Do not use the following reserved names for the name of a file:
>
> CON, PRN, AUX, NUL, COM1, COM2, COM3, COM4, COM5, COM6, COM7, COM8, COM9, LPT1, LPT2, LPT3, LPT4, LPT5, LPT6, LPT7, LPT8, and LPT9. Also avoid these names followed immediately by an extension; for example, NUL.txt is not recommended.

As far as I understand it these are reserved names and simply cannot be used as actual file names on Windows. They have a special meaning (backwards compatibility FTW) and will not (ever?) behave the way you want them to, i.e. as actual files.

**Comment by dasa** [comment]
I've run into a similar issue with cloning a repo that has a branch named `x/con`:

>error: cannot lock ref 'refs/remotes/origin/x/con': Unable to create 'C:/Users/jenkins/.jenkins/workspace/online/.git/refs/remotes/origin/x/con.lock': Invalid argument

Would it be possible for git to avoid this?

**Comment by dscho** [comment]
> Would it be possible for git to avoid this?

Not at this time. Git initially stores all refs as "loose refs", i.e. as files in the file system, where the path is relative to the Git directory.

**Comment by kant2002** [comment]
I think it's time to relax some Windows constraints, since on Windows 11 I was able to create file `prn.ap`, create file `con.lock`. Creation of `con` file is still work and I can see content of that file.

```
PS D:\d\github\aps> echo "123" >con.lock
PS D:\d\github\aps> more .\con.lock
123
PS D:\d\github\aps> echo "123" >con
PS D:\d\github\aps> more .\con
123
PS D:\d\github\aps> more con
Impossible d’accéder au fichier \\.\con
```

so probably all these shenanigans around filenames can be relaxed.

**Comment by dscho** [comment]
> I think it's time to relax some Windows constraints, since on Windows 11 I was able to create file `prn.ap`, create file `con.lock`.

Hmm. I do not think that it'd be wise to revise the checks that apply to _every_ Git for Windows installation solely based on _one particular_ setup... These checks exist to prevent such file names from being used in projects that need to check out on Windows (_all_ supported versions, that is).

**Comment by kant2002** [comment]
I read whole discussion and I agree with yoh position demanding more evidence that it is safe to introduce to more people. 

I see two routes how it can be hypothetically done
1. Create switch for Windows which relax current behavior or make it Win11 only. 
2. Research what Ms improve in the Win32 regarding this, maybe try to reach deep into Windows division and find people who can definitely say, or find people who can clarify documentation without exposing developers.

I will definitely try to go route 2, but it is quite possible that I would fail with that. 

I believe current position even if practical a bit paranoid in a sense that it overreacting when there evidence that it would work. 

If you see other ways forward, please let me know. If that in my power I would try something.

**Comment by mfide** [comment]
Reviving this with some data, since @kant2002 brought it up again.

## 1. Microsoft has acknowledged the change in writing, just not on the page we all read

[File path formats on Windows systems](https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats#handle-legacy-devices), section *Path normalization → Handle legacy devices*:

> **Prior to Windows 11**, a path that begins with a legacy device name is always interpreted as a legacy device by the `Path.GetFullPath(String)` method. For example, the DOS device path for `CON.TXT` is `\\.\CON`, and the DOS device path for `COM1.TXT\file1.txt` is `\\.\COM1`. **Because this no longer applies with Windows 11**, specify the full path to the legacy DOS device, such as `\\.\CON`.

Provenance, since "some page on learn.microsoft.com says so" isn't worth much on its own:

* added in [dotnet/docs#42370](https://github.com/dotnet/docs/pull/42370), merged 2024-08-29 by a Microsoft docs maintainer;
* it closes [dotnet/docs#41193](https://github.com/dotnet/docs/issues/41193), which reported that `Path.GetFullPath("CON.TXT")` now returns `C:\Users\<user>\CON.TXT`;
* the diff is literally `A path that begins with…` → `Prior to Windows 11, a path that begins with…`.

`Path.GetFullPath` is a thin wrapper over `GetFullPathNameW`, so this documents a change in Win32 path normalization, not a .NET-specific behaviour.

The Win32 page, [Naming Files, Paths, and Namespaces](https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file), has *not* been updated and still says *"NUL.txt and NUL.tar.gz are both equivalent to NUL"*. The two pages contradict each other today, and the Win32 one is the one everybody cites. Getting it corrected looks like the concrete deliverable of @kant2002's "route 2".

The new rules were also described independently by ChrisDenton and eryksun in [python/cpython#95486](https://github.com/python/cpython/issues/95486): a DOS device name is no longer special-cased if it has an extension, nor when it is the leaf component of a fully qualified path, with `NUL` as the sole exception.

## 2. Measured behaviour

Windows 11 Pro, build 26200 (25H2), local NTFS. Raw `CreateFileW` + `GetFullPathNameW` via P/Invoke, with no `\\?\` prefix, no shell, no .NET path helpers:

| input | `GetFullPathNameW` returns | real file created? |
|---|---|---|
| `aux.c` (relative) | `<dir>\aux.c` | yes |
| `aux.h` (relative) | `<dir>\aux.h` | yes |
| `con.lock` (relative) | `<dir>\con.lock` | yes |
| `nul.txt` (relative) | `<dir>\nul.txt` | yes |
| `lpt1.txt` (relative) | `<dir>\lpt1.txt` | yes |
| `aux` (relative) | `\\.\aux` | no (device) |
| `con` (relative) | `\\.\con` | no (device) |
| `com1` (relative) | `\\.\com1` | no (device) |
| `nul` (relative) | `\\.\nul` | no (device) |
| `<dir>\aux` (absolute) | `<dir>\aux` | **yes** |
| `<dir>\con` (absolute) | `<dir>\con` | **yes** |
| `<dir>\com1` (absolute) | `<dir>\com1` | **yes** |
| `<dir>\nul` (absolute) | `\\.\nul` | no (device) |

So: extension present → always an ordinary file; bare device name → still the device when the path is relative, but an ordinary file when fully qualified. `NUL` is the only unconditional exception left.

One caveat worth carrying forward: eryksun notes the SMB redirector still reserves these names in the filename part of an open/create, so working trees on network shares are not covered by any of this.

## 3. On this build, the only remaining blocker is Git's own check

`git version 2.52.0.windows.1`, local NTFS, files created out-of-band under the exact paths from the original report.

With defaults (`core.protectNTFS` unset, i.e. true):

```
$ git status --porcelain
?? README
?? drivers/

$ git add -A
error: open("drivers/gpu/drm/nouveau/nvkm/subdev/i2c/aux.c"): No such file or directory
error: unable to index file 'drivers/gpu/drm/nouveau/nvkm/subdev/i2c/aux.c'
fatal: adding files failed
```

The files exist and are readable; `git status` even lists them. The `ENOENT` is synthesised by `is_valid_win32_path()`, it is not what the OS returned.

With `core.protectNTFS=false`:

```
$ git add -A && git commit -m t
$ git ls-files
README
drivers/gpu/drm/nouveau/nvkm/subdev/i2c/aux.c
drivers/gpu/drm/nouveau/nvkm/subdev/i2c/aux.h

$ rm -rf drivers README
$ git checkout -f -- .          # exit 0
$ git status --porcelain        # clean
$ ls drivers/gpu/drm/nouveau/nvkm/subdev/i2c/
aux.c  aux.h

$ git branch x/con              # exit 0
$ ls .git/refs/heads/x/
con
```

That covers both the original report and @dasa's loose-ref case. Worth noting: when I filed this in 2022, the very same `protectNTFS=false` attempt failed with `unable to create file ...: No such file or directory`. That was Windows 10, where `CreateFile` genuinely was the blocker. On Windows 11 it no longer is.

## 4. What I think follows

@dscho, I don't read any of this as an argument for changing the default, and your "one particular setup" objection was fair. But the question is no longer whether it happens to work on somebody's machine: the behaviour change is documented by Microsoft, and the failure above is produced by Git, not by Windows.

Your concerns still hold for:

* network shares, where the SMB redirector is unchanged;
* `NUL`, which is still special-cased unconditionally;
* Windows versions that predate the change and remain supported: Server 2019 and 2022 run the older path normalization as far as I know (I have not tested those, and I'd be glad to if it helps).

Which is exactly why an opt-in escape hatch seems like the right shape, rather than a default change or the path-rewriting scheme you correctly said would break build systems: something like `core.protectWindowsReservedNames` (default `true`), in the same spirit as `core.protectNTFS`, letting people who know their environment clone the kernel without relaxing anything for anyone else.

Happy to run whatever further test matrix would be useful: other builds, Server SKUs, SMB shares, `\\?\` vs normalized paths, or the `FILE_FLAG_*` variations.

**Comment by dscho** [comment]
@mfide personally, I do like human-written summaries. Mostly because the economic force of having to type all the words yourself provides a natural constraint in the output length that benefits readers _a lot_. In other words, I find it really hard to distill the relevant information from this overly verbose, obviously AI-generated text, so I won't burden myself with that distillation task.

**Comment by mfide** [comment]
Short version: Microsoft changed the behavior and documented it. The .NET path documentation now says "Prior to Windows 11, a path that begins with a legacy device name is always interpreted as a legacy device ... Because this no longer applies with Windows 11": https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats#handle-legacy-devices

I did check the one thing your commit 4dc42c6c18 rests on, since that is what the ban is built on: on Windows 11, `CreateDirectoryW("NUL")` still returns success and still creates nothing, while `CON`, `AUX`, `PRN` and `COM1` fail with `ERROR_DIRECTORY`. So the silent-failure argument is still valid, but only for `NUL`.

**Comment by dscho** [comment]
> "Prior to Windows 11, [...]"

Well, we are _just about_ to drop support for Windows 8.1. So I think that all of this "but, but, but, but Windows 11!" objection to current Git for Windows' behavior is quite a bit premature.

**Comment by kant2002** [comment]
For the sake of discussion I propose this thing

```patch
diff --git a/compat/mingw.c b/compat/mingw.c
index fd014eef49..91d2af7928 100644
--- a/compat/mingw.c
+++ b/compat/mingw.c
@@ -167,6 +167,10 @@ int err_win_to_posix(DWORD winerr)
        return error;
 }

+int is_windows_11_or_greater(void) {
+       return GetVersion() > 22000 << 16;
+}
+
 static inline int is_file_in_use_error(DWORD errcode)
 {
        switch (errcode) {
@@ -4114,6 +4118,7 @@ int is_valid_win32_path(const char *path, int allow_literal_nul)
 segment_start:
                        switch (*path) {
                        case 'a': case 'A': /* AUX */
+                               if (is_windows_11_or_greater()) goto not_a_reserved_name;
                                if (((c = path[++i]) != 'u' && c != 'U') ||
                                    ((c = path[++i]) != 'x' && c != 'X')) {
 not_a_reserved_name:
@@ -4122,6 +4127,7 @@ int is_valid_win32_path(const char *path, int allow_literal_nul)
                                }
                                break;
                        case 'c': case 'C':
+                               if (is_windows_11_or_greater()) goto not_a_reserved_name;
                                /* COM1 ... COM9, CON, CONIN$, CONOUT$ */
                                if ((c = path[++i]) != 'o' && c != 'O')
                                        goto not_a_reserved_name;
@@ -4148,6 +4154,7 @@ int is_valid_win32_path(const char *path, int allow_literal_nul)
                                        goto not_a_reserved_name;
                                break;
                        case 'l': case 'L': /* LPT<N> */
+                               if (is_windows_11_or_greater()) goto not_a_reserved_name;
                                if (((c = path[++i]) != 'p' && c != 'P') ||
                                    ((c = path[++i]) != 't' && c != 'T') ||
                                    !isdigit(path[++i]))
@@ -4161,6 +4168,7 @@ int is_valid_win32_path(const char *path, int allow_literal_nul)
                                        goto not_a_reserved_name;
                                break;
                        case 'p': case 'P': /* PRN */
+                               if (is_windows_11_or_greater()) goto not_a_reserved_name;
                                if (((c = path[++i]) != 'r' && c != 'R') ||
                                    ((c = path[++i]) != 'n' && c != 'N'))
                                        goto not_a_reserved_name;
diff --git a/compat/mingw.h b/compat/mingw.h
index 379d7f5d07..11c84c767e 100644
--- a/compat/mingw.h
+++ b/compat/mingw.h
@@ -284,6 +284,11 @@ void open_in_gdb(void);
  */
 int err_win_to_posix(DWORD winerr);

+/**
+ * Checks if program run on Windows 11 or higher.
+ */
+int is_windows_11_or_greater(void);
+
 int mingw_platform_has_symlinks(void);
 #define platform_has_symlinks() mingw_platform_has_symlinks()



```

**Comment by kant2002** [comment]
Here the result of cloning. Looks not bad.

```
PS D:\d\github> git clone git://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git
Cloning into 'linux-stable'...
remote: Enumerating objects: 15038463, done.
remote: Counting objects: 100% (93574/93574), done.
remote: Compressing objects: 100% (33417/33417), done.
Receiving objects: 100% (15038463/15038463), 6.21 GiB | 577.00 KiB/s, done.
remote: Total 15038463 (delta 77411), reused 69724 (delta 60086), pack-reused 14944889 (from 1)
Resolving deltas: 100% (12012960/12012960), done.
Checking objects: 100% (33554432/33554432), done.
Updating files: 100% (96034/96034), done.
warning: the following paths have collided (e.g. case-sensitive paths
on a case-insensitive filesystem) and only one from the same
colliding group is in the working tree:

  'include/uapi/linux/netfilter/xt_CONNMARK.h'
  'include/uapi/linux/netfilter/xt_connmark.h'
  'include/uapi/linux/netfilter/xt_DSCP.h'
  'include/uapi/linux/netfilter/xt_dscp.h'
  'include/uapi/linux/netfilter/xt_MARK.h'
  'include/uapi/linux/netfilter/xt_mark.h'
  'include/uapi/linux/netfilter/xt_RATEEST.h'
  'include/uapi/linux/netfilter/xt_rateest.h'
  'include/uapi/linux/netfilter/xt_TCPMSS.h'
  'include/uapi/linux/netfilter/xt_tcpmss.h'
  'include/uapi/linux/netfilter_ipv4/ipt_ECN.h'
  'include/uapi/linux/netfilter_ipv4/ipt_ecn.h'
  'include/uapi/linux/netfilter_ipv4/ipt_TTL.h'
  'include/uapi/linux/netfilter_ipv4/ipt_ttl.h'
  'include/uapi/linux/netfilter_ipv6/ip6t_HL.h'
  'include/uapi/linux/netfilter_ipv6/ip6t_hl.h'
  'net/netfilter/xt_DSCP.c'
  'net/netfilter/xt_dscp.c'
  'net/netfilter/xt_HL.c'
  'net/netfilter/xt_hl.c'
  'net/netfilter/xt_RATEEST.c'
  'net/netfilter/xt_rateest.c'
  'net/netfilter/xt_TCPMSS.c'
  'net/netfilter/xt_tcpmss.c'
  'tools/memory-model/litmus-tests/Z6.0+pooncelock+poonceLock+pombonce.litmus'
  'tools/memory-model/litmus-tests/Z6.0+pooncelock+pooncelock+pombonce.litmus'
```

**Comment by dscho** [comment]
> ```diff
> +typedef void(WINAPI* pfnRtlGetNtVersionNumbers)(LPDWORD, LPDWORD, LPDWORD);
> +
> +int is_windows_11_or_greater(void) {
> +    pfnRtlGetNtVersionNumbers RtlGetNtVersionNumbers;
> +    DWORD major, minor, build;
> +
> +       HMODULE hNtdll = GetModuleHandleW(L"ntdll.dll");
> +    if (!hNtdll) return 0;
> +
> +#pragma GCC diagnostic push
> +#pragma GCC diagnostic ignored "-Wcast-function-type"
> +       RtlGetNtVersionNumbers = (pfnRtlGetNtVersionNumbers)GetProcAddress(hNtdll, "RtlGetNtVersionNumbers");
> +#pragma GCC diagnostic pop
> +
> +    if (!RtlGetNtVersionNumbers) return 0;
> +
> +    major = 0;
> +       minor = 0;
> +       build = 0;
> +    RtlGetNtVersionNumbers(&major, &minor, &build);
> +
> +    // Маска прибирає службові прапорці Microsoft з номери білду
> ```

s/Маска прибирає службові прапорці Microsoft з номери білду/Mask removes Microsoft service flags from the build number/

And if you look around in the source code, you will note the absence of C++-style comments: They are not allowed in Git's source code.

> ```diff
> +    build &= 0xFFFF;
> +
> +    // Windows 11 починається з Major 10 і Build >= 22000
> ```

s/починається з/begins with/

> ```diff
> +    return (major == 10 && build >= 22000);
> +}
> +
> +
> ```

What happened to using the simplest possible avenue, using `GetVersion()`? There's even [prior art](https://github.com/git-for-windows/git/blob/v2.55.0.windows.5/compat/mingw.c#L4359-L4371) in Git for Windows' source code for using that.

And still, this does not at all address the concern that I have now mentioned _multiple_ times, and therefore any patch that does not address this properly is premature: What about Windows 10 support? Just because _you_ can check out those files because you happen to have Windows 11 does not mean that other developers in the project are in the same position. It's like case-sensitivity: just because _your_ file system discerns between upper and lower case does not mean that your project _should_ contain files that conflict on case-insensitive file systems. And if those developers with a case-sensitive filesystem are not prevented by a Git error from using those conflicting file names, well, that's what you get, conflicting file names.

In other words: Just because you _can_ do it, you _can_ make such a patch, I don't think that it is a good idea to integrate it, not as long as Git for Windows supports older Windows versions. It would be paying mere lip service if we abandoned Windows 10 users this way.

**Comment by mfide** [comment]
Closing the loop on my original report, since I re-tested it this week.

The reserved-name problem is gone from the kernel itself :) Both files I hit in 2022 were renamed in 2024, with the same rationale in each commit message: "`the use of such file names is prohibited on other operating systems such as Windows, so the Linux repository cannot be  cloned and edited on them.`" (aux.c -> auxch.c)

So `core.protectNTFS=false` is only needed for Linux Kernel 6.12 and older. For anything current it makes no difference at all.

What is still needed on Windows is a case-sensitive directory, and that has nothing to do with this check.

I ran the `Linux Kernel v7.3-rc4` clone in a directory prepared with `fsutil file setCaseSensitiveInfo C:\linux enable` and it has all clean state.

So the practical recipe for the kernel today is: create the directory, run fsutil on it, clone into it. No git config needed unless you want a 6.12 or older branch.

As far as my original report goes, this is resolved on the kernel side. Whether the check itself should be narrowed is a separate question and I will leave it to the discussion above.

**Comment by kant2002** [comment]
@dscho I hear your concerns about Windows 10 and other OS pretty well. One of the reasons why I do not send PR is because I perfectly well understand that current patch is only partial implementation and you have specific concerns which I did not clarify. I don't know how to answer all of your concerns, but let's try to reduce number of them, maybe some other person in the future can. I assume that you don't take any patch unless you explicitly agree that this is a plan. This patch mostly for bystanders to simplify participation (if anybody cares).

`GetVersion` is indeed my oversight. I was a bit confused by lack of answer on the internet specifically regarding with Windows 11 check, and did not check myself. It seems it would do the trick.

Regarding Windows 10 users. I assume that we can leave existing behavior for OS which did not support CON/PRN/AUX/LPT on the kernel level. This protection make sense if kernel is not supported such names. So that's why I guard checks for Windows 11 specifically. If my check safe or not is separate question. I assume that we can make it safe and reliable.
